src/lib/optimizer/strategy/semi_join_reduction_rule.hpp

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "abstract_rule.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+class PredicateNode;
+
+/**
+ * A semi join reduction is a semi join that is added to the LQP without being present in the unoptimized LQP. This
+ * means that it does not change the final results. Take the following query as an example - it is loosely based on
+ * TPC-H 20:
+ *
+ *   SELECT p_name FROM part p1 WHERE p_size >
+ *     (SELECT AVG(p_size) FROM part p2 WHERE p1.p_container = p2.p_container)
+ *   AND p1.p_container IN ('SM CASE', 'MD CASE', 'LG CASE')
```

Comment:
Before digging into the code. What is the difference here between applying all predicates on both sides when ever they are connected via an equi-join? Which in something your other "in optimization" rule would reformulate to semi joins?

Reply:
Summary of our discussion, to be expanded in a comment:

A different approach would be to "copy" the predicates on the one side of an equi join to the other, as described in the Boncz paper. For that, we would want the information on how a table was filtered - extracting this from the LQP, especially across multiple predicates and joins, is difficult. While simple predicates might be cheaper than the semi join, the semi join approach has the advantage of being applicable no matter how the other side was constructed.

Reply:
Sounds good. Maybe something like "the semi join reduction implicitely applieds all predictes of one side on the other side"?

Reply:
Another point you handle but "predicate copying" does not is that you can exploit previous joins which removed many tuples.

Reply:
Also, predicates might be based on projections **and/or joined columns**, which makes propagation even more complex.